### PR TITLE
Issue #27 - Support extension OID 1.3.6.1.4.1.311.21.2

### PR DESCRIPTION
**Pull request by MauAraujo**
Hey @april, I finally implemented the support for the extension. However, I am having a hard time finding a page that uses this certificate extension, I used Censys to search for one but had no results. So, I still haven't tested it completely. Any advice and feedback would be greatly appreciated.

### CODE REVIEW

File: src/viewer/js/der.js [hunk 1]
```diff
@@ -276,6 +277,17 @@ export const parse = async (der) => {
     timestamps: scts,
   }
 
+    console.log(x509.extensions);
+  // get the Microsoft certificate serv
+  let mcsrv = {
+      previousHash: getX509Ext(x509.extensions, '1.3.6.1.4.1.311.21.2').parsedValue,
```
Comment (april) [inline src/viewer/js/der.js]: `parsedValue` is an object, not a string, and it is dependent on PKIjs natively supporting it (which I don't believe it does).

File: src/viewer/js/strings.js [hunk 2]
```diff
@@ -321,6 +321,12 @@ export const strings = {
         long: 'Extended Key Usages',
       }
     },
+    '1.3.6.1.4.1.311.21.2': {
+       name: {
+        short: 'MS Cert Server Prev',
```
Comment (april) [inline src/viewer/js/strings.js]: This is a really odd short name, it should probably just be Previous Hash.

File: src/viewer/js/der.js [hunk 3]
```diff
@@ -276,6 +277,17 @@ export const parse = async (der) => {
...
+  }
+  if(mcsrv.previousHash) {
```
Comment (april) [inline src/viewer/js/der.js]: Spacing issues here.

File: src/viewer/index.handlebars [hunk 4]
```diff
@@ -342,6 +342,21 @@
...
+            {{! Microsoft Certificate Server }}
+              <div class="panel-section-subheader">
+                <span>
+                {{#if this.ext.mcsrv.critical}}<img class="critical" alt="critical extension" title="This extension has been marked as critical, meaning that clients must reject the certificate if they don't understand it." src="../icons/critical.svg">{{/if}}
+                Certificate Server
+                </span>
+              </div>
+              <div class="panel-list-item">
+                <div class="text">Previous Hash</div>
+                <div class="text breakable">{{this.ext.mcsrv.previousHash}}</div>
```
Comment (april) [inline src/viewer/index.handlebars]: Please fix the spacing to be like the rest of the template.  :)

File: src/viewer/js/der.js [hunk 5]
```diff
@@ -61,6 +61,7 @@ export const parse = async (der) => {
     '1.3.6.1.4.1.11129.2.4.2',  // embedded scts
     '1.3.6.1.5.5.7.1.1',        // authority info access
     '1.3.6.1.5.5.7.1.24',       // ocsp stapling
+    '1.3.6.1.4.1.311.21.2',     // Microsoft Cert Server previous
```
Comment (april) [inline src/viewer/js/der.js]: Previous hash?

File: src/viewer/js/der.js [hunk 6]
```diff
@@ -276,6 +277,17 @@ export const parse = async (der) => {
     timestamps: scts,
   }
 
+    console.log(x509.extensions);
```
Comment (april) [inline src/viewer/js/der.js]: Please do not leave console logging in your commits.

**Comment by april** [comment]
Right at the top of `der.js`, you can set `asn1` to the value of a decoded cert that does have this value. You should be able to grab the certs off of Censys or the like.

**Comment by MauAraujo** [comment]
Hi, I tried setting `asn1` manually as suggested but ran into some issues. I tried to manually read a certificate that I downloaded from Censys, but when creating asn1: 

```javascript
  const asn1 = asn1js.fromBER(testCert.buffer);
```
I keep getting the error: 
```javascript
"Too big integer"
```

`testCert ` is an ArrayBuffer that contains the certificate I want to decode. I would really appreciate any insight into this problem. Thanks.

**Comment by MauAraujo** [comment]
Nevermind, I was able to decode it using another approach. 
I decoded this certificate https://censys.io/certificates/c11d158f29a7d8c8eff057550fddd3a4a6b97119bc62e60bb8680680b05397e7 that contains the extension, and it displays correctly: 
Captura de pantalla 2019-03-18 a la(s) 17 53 51
The changes I submitted previously are shown here:
Captura de pantalla 2019-03-18 a la(s) 17 56 16
However, upon further examination, the extension does not contain a `parsedValue`so the value cannot be displayed.  This is the structure of the extension: 

```javascript
Object { extnID: "1.3.6.1.4.1.311.21.2", extnValue: {…} }
```
Do I have to decode the value manually or is there anything I'm missing? Thanks for the support and feedback.

**Comment by april** [comment]
Yes, if PKI.js can't parse it you can either open up an Issue/PR there (in their GitHub) or try to parse it inside the extension.  :)

**Comment by MauAraujo** [comment]
I used the new version of PKI.js that supports the Microsoft extensions and was able to get `parsedValue` and display the previous hash id like shown here:
  
prevhash

`parsedValue` has this structure: 
```javascript
parsedValue: {…}
​​
blockLength: 22
​​
blockName: "OctetString"
​​
error: ""
​​
idBlock: Object { blockName: "identificationBlock", blockLength: 1, isHexOnly: false, … }
​​
lenBlock: Object { blockName: "lengthBlock", blockLength: 1, isIndefiniteForm: false, … }
​​
valueBeforeDecode: "041430E37E3720456C38C8874F382D07CE23AA2303C1"
​​
valueBlock: Object { blockName: "OctetStringValueBlock", blockLength: 20, isIndefiniteForm: false, … }
​​
warnings: Array []
​​
<prototype>: Object { … }
```
Is this correct? Or is there another field that I need to consider?

**Comment by MauAraujo** [comment]
This was changed and included in #54.

**Pull request closed without merging**